entropylab_qpudb/_quaconfig.py

```diff
@@ -133,18 +142,52 @@ def update_intermediate_frequency(
 
     def update_op_amp(self, element, operation, amp):
         pulse = self.get_pulse_from_op(element, operation)
-        try:
-            self.data["waveforms"][pulse["waveforms"]["single"]]["sample"] = amp
-        except KeyError:
+        if "sample" not in self.data["waveforms"][pulse["waveforms"]["single"]]:
             raise KeyError(
                 "Can only access amplitude for a constant pulse of a single element"
             )
+        else:
+            self.data["waveforms"][pulse["waveforms"]["single"]]["sample"] = amp
 
     def get_op_amp(self, element, operation):
+        """
+        DEPRECATED - DO NOT USE
+        use `get_waveforms_from_op` instead
+        """
         pulse = self.get_pulse_from_op(element, operation)
         try:
             return self.data["waveforms"][pulse["waveforms"]["single"]]["sample"]
         except KeyError:
             raise KeyError(
                 "Can only access amplitude for a constant pulse of a single element"
             )
+
+    def get_port_by_element(self, element: str, port: str) -> Tuple[str, int]:
+        """
+        returns the ports of a quantum element.
+        :param element: Name of the element
+        :param port: Name of the port. Can be either 'single' if element has `singleInput` or 'I', 'Q' if element has mixed inputs.
+        :return: a tuple of the form (con_name, port number)
+        """
+        element_data = self.data["elements"][element]
+        if port == 'single':
+            if "singleInput" in element_data:
+                return element_data["singleInput"]["port"]
+            else:
+                raise ValueError("can only use 'single' for a singleInput quantum element")
+        if "mixInputs" in element_data:
+            if port == "I" or port == "Q":
+                return element_data["mixInputs"][port]
+            else:
+                raise ValueError("can only use 'I' or 'Q' for a mixInputs quantum element")
+
+    def set_output_dc_offset_by_element(self, element: str, port: str, offset: float) -> None:
+        """
+        Set a DC offset value by element
+
+        :param element: Name of the element
+        :param port: Name of the port. Can be either 'single' if element has `singleInput` or 'I', 'Q' if element has mixed inputs.
+        :param offset: offset value to set
+        """
+        port = self.get_port_by_element(element, port)
+        self.data["controllers"][port[0]]["analog_outputs"][port[1]]["offset"] = offset
```

Comment:
is this correct? seems weird that port is under controllers and also under analog input

Reply:
you mean that I use the term "port" for both the input to the function (i.e. the input of the element) and for the actual port? You're right, it's weird.
I think the mapping is "get controller port from element input" - this is what makes sense. I'll change it and you can review again